internal/releasesources/release_sources.go

```diff
@@ -0,0 +1,510 @@
+	// Currently OCM_BASE_URL=https://api.<something>openshift.com such that <something>=(integration.)/(stage.)/(.),
+	// which means scheme and host
+	ocmBaseURL, err := url.Parse(config.OCMBaseURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error occurred while trying to parse OCM base URL: %s", config.OCMBaseURL)
+	}
+	ocmBaseURL.Path = openshiftUpdateServiceAPIURLPath
+	// Currently RED_HAT_PRODUCT_LIFE_CYCLE_DATA_API_BASE_URL=https://access.redhat.com/product-life-cycles/api/v1/products,
+	// which means schema, host and path
+	redHatCustomerPortalBaseURL, err := url.Parse(config.RedHatProductLifeCycleDataAPIBaseURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error occurred while trying to parse Red Hat Customer Portal base URL: %s", config.RedHatProductLifeCycleDataAPIBaseURL)
+	}
+
+	return &releaseSourcesHandler{
+		releaseSources:      releaseSources,
+		staticReleaseImages: releaseImages,
+		log:                 logger,
+		db:                  db,
+		config:              config,
+		releaseHandler:      releaseHandler,
+		releasesClient: openShiftReleasesAPIClient{
+			baseURL: *ocmBaseURL,
+		},
+		supportLevelClient: openShiftSupportLevelAPIClient{
+			baseURL: *redHatCustomerPortalBaseURL,
+		},
+	}, nil
+}
+
+func (h *releaseSourcesHandler) createReleaseImage(channel, version, cpuArchitecture string) (*common.ReleaseImage, error) {
+	majorMinorVersion, err := common.GetMajorMinorVersion(version)
+	if err != nil {
+		return nil, err
+	}
+
+	cpuArchitectures := []string{cpuArchitecture}
+	newReleaseImageReference := getReleaseImageReference(version, cpuArchitecture)
+
+	newOpenshiftVersion := majorMinorVersion
+	newVersion := &version
+	if cpuArchitecture == common.MultiCPUArchitecture {
+		newOpenshiftVersion = swag.String(fmt.Sprintf("%s-%s", *newOpenshiftVersion, common.MultiCPUArchitecture))
+		newVersion = swag.String(fmt.Sprintf("%s-%s", *newVersion, common.MultiCPUArchitecture))
+		cpuArchitectures, err = h.releaseHandler.GetReleaseArchitecture(h.log, newReleaseImageReference, "", "")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get the CPU architecturs of the release image with reference: %s", newReleaseImageReference)
+		}
+	}
+
+	return &common.ReleaseImage{
+		ReleaseImage: models.ReleaseImage{
+			OpenshiftVersion: newOpenshiftVersion,
+			CPUArchitecture:  &cpuArchitecture,
+			Version:          newVersion,
+			URL:              swag.String(newReleaseImageReference),
+			Default:          false,
+		},
+		Channel:            &channel,
+		DBCPUArchitectures: cpuArchitectures,
+	}, nil
+}
+
+func (h *releaseSourcesHandler) insertToDB(tx *gorm.DB, dbReleaseImages []*common.ReleaseImage) error {
+	h.log.Debugf("Inserting %d records to openshift_versions table", len(dbReleaseImages))
+
+	if len(dbReleaseImages) == 0 {
+		return nil
+	}
+
+	return tx.Create(dbReleaseImages).Error
+}
+
+// removeReleaseImageDuplicates removes release images such that only one release image
```

Comment:
Why we will have dumplicates?

Reply:
First, most releases exist in more than one channel, so if we collect for example 4.14 releases from `stable` and `candidate`, we will have the same release twice. Second I thought it might be possible for the API to mistake and give me duplicate records, which will cause us an error as the `url` (whic his composed of `version` and `arch`) is the primary key.